PCF-ADR-0007-operator-redesign.md

```diff
@@ -0,0 +1,201 @@
+	BackoffTime int `json:"backoffTime,omitempty"`
+	// Head of line job will not be bumped away for at least HeadOfLineHoldingTime seconds
+	// by higher priority jobs.
+	// Default setting to 0 disables this mechanism.
+	HeadOfLineHoldingTime int `json:"headOfLineHoldingTime,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+InstaScale would also declare its configuration, e.g.:
+
+```go
+package config
+
+struct InstaScaleConfiguration {
+	// Maximum of cluster nodes that can be scaled out.
+	// Defaults to 3.
+	MaxScaleoutAllowed int `json:"maxScaleoutAllowed,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+Finally, the CodeFlare operator would embed MCAD and InstaScale configurations, along with the generic configuration taken from [component-base](https://github.com/kubernetes/component-base), e.g.:
+
+```go
+package config
+
+import (
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
+	mcad "github.com/project-codeflare/multi-cluster-app-dispatcher/pkg/config"
+	instascale "github.com/project-codeflare/instascale/pkg/config"
+)
+
+struct InstaScaleConfiguration {
+	// Enabled controls whether the InstaScale controller is started.
+	// It may defaults to true on platforms that InstaScale supports.
+	// Otherwise defaults to false.
+	Enabled *bool `json:"enabled,omitempty"`
+
+	// The InstaScale controller configuration
+	instascale.InstaScaleConfiguration `json:",inline,omitempty"`
+}
+
+struct CodeFlareOperatorConfiguration {
+	// The MCAD controller configuration
+	MCAD *mcad.MCADConfiguration `json:"mcad,omitempty"`
+
+	// The InstaScale controller configuration
+	InstaScale *InstaScaleConfiguration `json:"instascale,omitempty"`
+
+	// ClientConnection configures the connection to the cluster
+	ClientConnection *configv1alpha1.ClientConnectionConfiguration `json:"clientConnection,omitempty"`
+
+	// LeaderElection is the LeaderElection config to be used when configuring
+	// the manager.Manager leader election
+	LeaderElection *configv1alpha1.LeaderElectionConfiguration `json:"leaderElection,omitempty"`
+
+	// Remaining configuration...
+}
+```
+
+The later `struct` would be sourced from a ConfigMap, mounted into the operator deployment container.
+To follows the Operator SDK best practice, which states an operator [should always be able to deploy and come up without user input](https://sdk.operatorframework.io/docs/best-practices/best-practices/#summary-1), that ConfigMap would be created when the operator starts, if it doesn't exist already, and initialized with sensible defaults.
+
+We'll provide documentation for existing users, on how to migrate their changes in the MCAD and InstaScale resources, over to the new ConfigMap resource.
+
+### Packaging and Automation
+
+The building and publishing of the MCAD and InstaScale container images can be removed.
+The CodeFlare operator will depend on the MCAD and InstaScale controllers Go packages,
+so only tagging their respective repositories will be needed.
```

Comment:
So it means that if someone needs to test own change i.e. in MCAD then the whole operator needs to be built, right?

Reply:
Right, the operator container image has to be rebuild, instead of the MCAD container image currently.

Reply:
Thinking about testing now....
MCAD has e2e kuttl tests using KinD, these tests would either need to be migrated to test go code directly, or need to build the whole operator.

Reply:
I think existing e2e tests could be left as is, and apply what's described in the _Local Development_ section, i.e., MCAD could still be deployed using Helm locally. That being said, starting the `main` process locally would also work.